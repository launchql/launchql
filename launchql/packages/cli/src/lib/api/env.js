import walkup from 'node-walkup';
import { join } from 'path';
import { readFileSync, readdirSync, writeFileSync } from 'fs';
const CONFIG_JSON = 'config.json';

// export const lqlPath = (cwd = process.cwd()) => {
//   let obj;
//   return new Promise((resolve, reject) => {
//     // if (process.env.LAUNCHQL_PATH) {
//     //   return resolve(process.env.LAUNCHQL_PATH);
//     // }
//     if (obj) {
//       return resolve(obj);
//     }
//     walkup(
//       '.lqlrc',
//       {
//         cwd
//       },
//       (err, matches) => {
//         if (err) {
//           return reject(err);
//         }
//         if (!matches || !matches.length) {
//           console.error('Not inside of a LaunchQL project');
//         }
//         obj = join(matches[0].dir, matches[0].files[0]);
//         resolve(obj);
//       }
//     );
//   });
// };

export const getLqlDir = (cwd = process.cwd()) => {
  let obj;
  return new Promise((resolve, reject) => {
    if (obj) {
      return resolve(obj);
    }
    walkup(
      '.lql',
      {
        cwd
      },
      (err, matches) => {
        if (err) {
          return reject(err);
        }
        if (!matches || !matches.length) {
          console.error('cannot find a .lql directory anywhere');
        }
        obj = join(matches[0].dir, matches[0].files[0]);
        resolve(obj);
      }
    );
  });
};

// lqlEnv is for local environments, db info, etc

// export const lqlEnv = async (cwd = process.cwd()) => {
//   let env;
//   const path = await lqlPath(cwd);
//   return new Promise((resolve, reject) => {
//     if (env) {
//       return resolve(env);
//     }
//     env = {};
//     readFileSync(path)
//       .toString()
//       .split('\n')
//       .map((e) => e.trim())
//       .forEach((vari) => {
//         const line = vari.trim();
//         if (!line || !line.length || /^#/.test(line)) return;
//         vari.replace(/(\w+)=(.+)/g, function ($0, $1, $2) {
//           env[$1] = $2;
//         });
//         return vari;
//       });

//     resolve(env);
//   });
// };

// export const lqlContext = async (cwd = process.cwd()) => {
//   const env = await lqlEnv(cwd);
//   const config = await getConfig(cwd);
//   return {
//     env,
//     config
//   };
// };

export const setCurrentContext = async (name, cwd = process.cwd()) => {
  const config = await getConfig(cwd);
  config.currentContext = name;
  await setConfig(config, cwd);
};

export const setConfig = async (config, cwd = process.cwd()) => {
  const dir = await getLqlDir(cwd);
  writeFileSync(join(dir, CONFIG_JSON), JSON.stringify(config, null, 2));
};

export const getCurrentContext = async (cwd = process.cwd()) => {
  const config = await getConfig(cwd);

  if (!config) {
    console.warn('no config found!');
    return {};
  }

  const current = config.currentContext;
  const context = config.contexts?.find((context) => context.name == current);
  if (!context) {
    console.warn('no context found!');
    return {};
  }
  context.user = config.users?.find((el) => el.name == context.user);
  context.server = config.servers?.find((el) => el.name == context.server);
  return context;
};

// lqlPath is for local environments, db info, etc
// getConfig
export const getConfig = async (cwd = process.cwd()) => {
  // use metaphors from KUBECONFIG
  // for now we just get the directory recursively...
  // KUBECONFIG => LQLCONFIG

  const dir = await getLqlDir(cwd);
  const ls = readdirSync(dir);

  let config = {};
  if (ls.includes(CONFIG_JSON)) {
    config = JSON.parse(readFileSync(join(dir, CONFIG_JSON)).toString());
  } else {
    console.warn('missing lql config');
    console.log(ls);
    return {};
  }

  return config;
};
