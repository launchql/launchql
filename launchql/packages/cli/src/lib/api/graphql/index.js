// ==== Queries
export * from './queries/get-database';
export * from './queries/get-databases';
export * from './queries/get-schemata';
export * from './queries/get-tables';
export * from './queries/get-module-definitions';
export * from './queries/get-tables-by-visibility';
export * from './queries/get-table';
export * from './queries/get-all-constraints';
export * from './queries/get-constraints';
export * from './queries/get-foreign-key-constraints';
export * from './queries/get-primary-key-constraints';
export * from './queries/get-module-outputs-by-definition-ids';
export * from './queries/get-unique-constraints';
export * from './queries/get-sql-actions';
export * from './queries/get-services';

// ==== Mutations
export * from './mutations/create-database';
export * from './mutations/update-database';
export * from './mutations/delete-database';

export * from './mutations/create-module';

export * from './mutations/create-table';
export * from './mutations/update-table';
export * from './mutations/delete-table';

export * from './mutations/create-field';
export * from './mutations/update-field';
export * from './mutations/delete-field';

export * from './mutations/create-foreign-key-constraint';
export * from './mutations/create-primary-key-constraint';
export * from './mutations/create-unique-constraint';
export * from './mutations/update-constraint';
export * from './mutations/delete-constraint';
