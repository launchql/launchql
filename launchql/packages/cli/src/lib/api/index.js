import { makeAutocompleteFunctionWithInput as makeSearch } from '@launchql/db-utils';

import createDatabase from './commands/create/database';
import createField from './commands/create/field';
import createTable from './commands/create/table';
import createModule from './commands/create/module';
import createRelation from './commands/create/relation';
import createUnique from './commands/create/unique';
import createPrimary from './commands/create/primary';

import updateField from './commands/update/field';
import updateTable from './commands/update/table';
import updateDatabase from './commands/update/database';

import deleteField from './commands/delete/field';
import deleteTable from './commands/delete/table';
import deleteDatabase from './commands/delete/database';

import exportDatabase from './commands/export/database';
import exportSdk from './commands/export/sdk';

import listDatabases from './commands/list/databases';
import listSchemata from './commands/list/schemata';
import listTables from './commands/list/tables';
import listFields from './commands/list/fields';

import configCreds from './commands/config/set-credentials';
import configServer from './commands/config/set-server';
import configContext from './commands/config/use-context';
import configView from './commands/config/view';
import configCreate from './commands/config/create';

// later make this generated from folders/files nested
export const functions = {
  config: {
    create: configCreate,
    user: configCreds,
    server: configServer,
    context: configContext,
    view: configView
  },
  create: {
    primaryKey: createPrimary,
    uniqueKey: createUnique,
    relation: createRelation,
    module: createModule,
    field: createField,
    table: createTable,
    database: createDatabase
  },
  delete: {
    field: deleteField,
    table: deleteTable,
    database: deleteDatabase
  },
  update: {
    field: updateField,
    database: updateDatabase,
    table: updateTable
  },
  export: {
    database: exportDatabase,
    sdk: exportSdk
  },
  list: {
    fields: listFields,
    tables: listTables,
    databases: listDatabases,
    schemata: listSchemata
  }
};

export const questions = Object.keys(functions).reduce((m, v) => {
  const srch = Object.keys(functions[v]);
  if (srch.length) {
    m[v] = [
      {
        _: true,
        type: 'autocomplete',
        name: 'action',
        message: 'which action?',
        source: makeSearch(srch)
      }
    ];
  }

  return m;
}, {});

export const objects = Object.keys(questions);
