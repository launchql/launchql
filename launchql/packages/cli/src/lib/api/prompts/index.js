import { prompt } from 'inquirerer';
import {
  getDatabasesQuery,
  getSchemataQuery,
  getTablesQuery,
  createDatabaseMutation,
  createTableMutation,
  createFieldMutation
} from '../graphql';

import { makeAutocompleteFunctionWithInput as makeSearch } from '@launchql/db-utils';

export const getDatabase = async (client, args) => {
  const result = await client.request(getDatabasesQuery);
  const database = await prompt(
    [
      {
        type: 'autocomplete',
        name: 'database',
        message: 'enter a database name',
        source: makeSearch(result.databases.nodes.map((n) => n.name)),
        required: true
      }
    ],
    args
  );
  return result.databases.nodes.reduce((m, v) => {
    if (v.name === database.database) return v;
    return m;
  });
};

export const getSchema = async (client, schemata, args) => {
  const result = await prompt(
    [
      {
        type: 'autocomplete',
        name: 'schema',
        message: 'enter a schema name',
        source: makeSearch(schemata.nodes.map((n) => n.name)),
        required: true
      }
    ],
    args
  );

  return schemata.nodes.reduce((m, v) => {
    if (v.name === result.schema) return v;
    return m;
  });
};

export const getTable = async (client, tables, args) => {
  const result = await prompt(
    [
      {
        type: 'autocomplete',
        name: 'table',
        message: 'enter a table name',
        source: makeSearch(tables.nodes.map((n) => n.name)),
        required: true
      }
    ],
    args
  );

  return tables.nodes.reduce((m, v) => {
    if (v.name === result.table) return v;
    return m;
  });
};

export const getAny = async (client, ctx, args) => {
  const { key, nodes, nameKey = 'name' } = ctx;
  const message = ctx.message || `enter value for ${key}`;
  const result = await prompt(
    [
      {
        type: 'autocomplete',
        name: key,
        message,
        source: makeSearch(nodes.map((n) => n[nameKey])),
        required: true
      }
    ],
    args
  );

  return nodes.reduce((m, v) => {
    if (v[nameKey] === result[key]) return v;
    return m;
  });
};

export const getField = async (client, fields, args) => {
  const result = await prompt(
    [
      {
        type: 'autocomplete',
        name: 'field',
        message: 'enter a field name',
        source: makeSearch(fields.nodes.map((n) => n.name)),
        required: true
      }
    ],
    args
  );

  return fields.nodes.reduce((m, v) => {
    if (v.name === result.field) return v;
    return m;
  });
};
