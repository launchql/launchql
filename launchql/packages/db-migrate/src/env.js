import { cleanEnv, str, port } from 'envalid';

export default cleanEnv(
  process.env,
  {
    PGUSER: str({ default: 'postgres' }),
    PGHOST: str({ default: 'localhost' }),
    PGPASSWORD: str({ default: 'password' }),
    PGDATABASE: str({ default: 'webinc-db' }),
    PGPORT: port({ default: 5432 })
  },
  { dotEnvPath: null }
);
