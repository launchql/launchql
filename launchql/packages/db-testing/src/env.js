import { cleanEnv, bool, port, str } from 'envalid';

export default cleanEnv(
  process.env,
  {
    PGTEMPLATE_DATABASE: str({ default: 'testing-template-db' }),
    PGHOST: str({ default: 'localhost' }),
    PGPASSWORD: str({ default: 'password' }),
    PGPORT: port({ default: 5432 }),
    PGUSER: str({ default: 'postgres' }),
    APP_USER: str({ default: 'app_user' }),
    APP_PASSWORD: str({ default: 'app_password' }),
    FAST_TEST: bool({ default: false })
  },
  { dotEnvPath: null }
);
