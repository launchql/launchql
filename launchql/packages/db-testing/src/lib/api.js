import { wrapConn } from './helpers';
import { getConnections as getConns } from './testing';

export const getApi = async ([pub, priv]) => {
  const { db, conn, teardown } = await getConns();
  const api = {
    public: wrapConn(conn, pub),
    private: wrapConn(conn, priv)
  };
  const admin = {
    public: wrapConn(db, pub),
    private: wrapConn(db, priv)
  };
  return { api, admin, db, conn, teardown };
};
