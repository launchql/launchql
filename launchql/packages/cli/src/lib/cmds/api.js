import { exec } from 'child_process';
import { prompt } from 'inquirerer';
import { makeAutocompleteFunctionWithInput } from '@launchql/db-utils';

import { objects, questions, functions } from '../api';
import { getCurrentContext } from '../api/env';

import { GraphQLClient } from 'graphql-request';

const search = makeAutocompleteFunctionWithInput(objects);

const question = [
  {
    _: true,
    type: 'autocomplete',
    name: 'object',
    message: 'which object type?',
    source: search
  }
];

export const aliases = ['a'];
export default async (argv) => {
  const { object } = await prompt(question, argv);
  const env = await getCurrentContext();
  let context;
  if (env && env.server) {
    const headers = {};
    if (env.user.hasOwnProperty('auth')) {
      if (env.user.auth.type === 'basic') {
        let str = `${env.user.auth.username}:${env.user.auth.password}`;
        if (
          typeof window !== 'undefined' &&
          typeof window.btoa === 'function'
        ) {
          str = window.btoa(str);
        } else {
          str = Buffer.from(str).toString('base64');
        }
        headers.authorization = `Basic ${str}`;
      } else if (env.user.auth.type === 'token') {
        headers.authorization = `Bearer ${env.user.auth.accessToken}`;
      }
    }

    const db = new GraphQLClient(
      env.server.endpoints?.db || env.server.endpoint,
      { headers }
    );
    const migrate = new GraphQLClient(
      env.server.endpoints?.migrate || env.server.endpoint,
      { headers }
    );
    const svc = new GraphQLClient(
      env.server.endpoints?.svc || env.server.endpoint,
      { headers }
    );
    const mods = new GraphQLClient(
      env.server.endpoints?.mods || env.server.endpoint,
      { headers }
    );
    context = {
      env,
      db,
      migrate,
      svc,
      mods
    };
  }

  if (!questions[object]) throw new Error(`cannot find ${object}`);
  const { action } = await prompt(questions[object], argv);

  if (typeof functions[object][action] !== 'function') {
    throw new Error(`${action} not found!`);
  }
  try {
    await functions[object][action](context, argv);
  } catch (e) {
    console.error('Error:');
    console.error(e.message);
  }
};
