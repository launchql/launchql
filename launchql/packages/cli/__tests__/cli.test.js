import { getCurrentContext, getConfig } from '../src/lib/api/env';

const LQLPATH = __dirname + '/../__fixtures__/';

let env, context, config, current;
beforeAll(async () => {
  config = await getConfig(LQLPATH);
  current = await getCurrentContext(LQLPATH);
});

describe('configurations', () => {
  it('config', async () => {
    expect(config).toMatchSnapshot();
  });
  it('current', async () => {
    expect(current).toMatchSnapshot();
  });
});
