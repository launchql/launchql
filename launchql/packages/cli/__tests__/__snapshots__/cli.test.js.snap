// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configurations config 1`] = `
Object {
  "apiVersion": "v1",
  "contexts": Array [
    Object {
      "name": "dans-laptop",
      "server": "local",
      "user": "admin",
    },
    Object {
      "name": "production",
      "server": "prod",
      "user": "pyramation",
    },
  ],
  "currentContext": "dans-laptop",
  "servers": Array [
    Object {
      "endpoints": Object {
        "db": "http://collection.lql.localhost:5555/graphql",
        "migrate": "http://migrate.lql.localhost:5555/graphql",
        "mods": "http://mods.lql.localhost:5555/graphql",
        "svc": "http://svc.lql.localhost:5555/graphql",
      },
      "env": "development",
      "name": "local",
    },
    Object {
      "endpoint": "https://admin.io/graphql",
      "env": "production",
      "name": "prod",
    },
  ],
  "users": Array [
    Object {
      "name": "pyramation",
      "token": Object {
        "accessToken": "redacted",
        "refreshToken": "redacted",
      },
    },
    Object {
      "basic": Object {
        "header": "randomstuffgoeshere",
        "password": "redactedtillweneedit",
        "username": "admin",
      },
      "name": "admin",
    },
    Object {
      "basic": Object {
        "header": "randomstuffgoeshere",
        "password": "redactedtillweneedit",
        "username": "user",
      },
      "name": "user",
    },
  ],
}
`;

exports[`configurations current 1`] = `
Object {
  "name": "dans-laptop",
  "server": Object {
    "endpoints": Object {
      "db": "http://collection.lql.localhost:5555/graphql",
      "migrate": "http://migrate.lql.localhost:5555/graphql",
      "mods": "http://mods.lql.localhost:5555/graphql",
      "svc": "http://svc.lql.localhost:5555/graphql",
    },
    "env": "development",
    "name": "local",
  },
  "user": Object {
    "basic": Object {
      "header": "randomstuffgoeshere",
      "password": "redactedtillweneedit",
      "username": "admin",
    },
    "name": "admin",
  },
}
`;
